reasoner/planner/AnswerCountEstimator.java

```diff
@@ -60,190 +59,76 @@
-            this.cyclicConcludables = cyclicConcludables;
-            this.acyclicConcludables = iterate(resolvables)
-                    .filter(resolvable -> resolvable.isConcludable() && !cyclicConcludables.contains(resolvable.asConcludable()))
-                    .map(Resolvable::asConcludable).toSet();
-        }
-
-        public FunctionalIterator<Resolvable<?>> resolvables() {
-            return iterate(resolvables);
-        }
-
-        public FunctionalIterator<Retrievable> retrievables() {
-            return iterate(resolvables).filter(Resolvable::isRetrievable).map(Resolvable::asRetrievable);
-        }
-
-        public FunctionalIterator<Negated> negateds() {
-            return iterate(resolvables).filter(Resolvable::isNegated).map(Resolvable::asNegated);
-        }
-
-        public FunctionalIterator<Concludable> cyclicConcludables() {
-            return iterate(cyclicConcludables);
-        }
-
-        public FunctionalIterator<Concludable> acyclicConcludables() {
-            return iterate(acyclicConcludables);
-        }
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variables, Set<Resolvable<?>> resolvables) {
+        if (!conjunctionModels.containsKey(conjunction)) buildConjunctionModel(conjunction);
+        Set<Variable> estimateableVariables = ReasonerPlanner.estimateableVariables(variables);
+        assert conjunction.pattern().variables().containsAll(estimateableVariables);
+        assert logicMgr.compile(conjunction).containsAll(resolvables);
+        return conjunctionModels.get(conjunction).estimateAnswers(estimateableVariables, resolvables);
     }
 
     private static class ConjunctionModel {
-        private final ConjunctionContext conjunctionContext;
+        private final ConjunctionSummary conjunctionSummary;
         private final Map<Variable, LocalModel.VariableModel> variableModels;
         private final Map<Resolvable<?>, List<LocalModel>> constraintModels;
         private final boolean isCyclic;
 
-        private ConjunctionModel(ConjunctionContext conjunctionContext,
+        private ConjunctionModel(ConjunctionSummary conjunctionSummary,
                                  Map<Variable, LocalModel.VariableModel> variableModels, Map<Resolvable<?>, List<LocalModel>> constraintModels,
                                  boolean isCyclic) {
-            this.conjunctionContext = conjunctionContext;
+            this.conjunctionSummary = conjunctionSummary;
             this.variableModels = variableModels;
             this.constraintModels = constraintModels;
             this.isCyclic = isCyclic;
         }
 
         private long estimateAnswers(Set<Variable> variables, Set<Resolvable<?>> resolvables) {
+            if (resolvables.size() ==0 ) return 1; // TODO: Is this right, or should we use the type-upper-bound for the unconstrained variables?
```

Comment:
make sure you run auto-format and import-cleanup on the changed classes as a habit

Reply:
why is it possible to have an empty set of resolvables at all? Same as having an empty set of variables?